### PR TITLE
gh-112175: Add `eval_breaker` to `PyThreadState`

### PR DESCRIPTION
**Pull request by swtaarrs**
This change adds an `eval_breaker` field to `PyThreadState`, renaming the existing `PyInterpreterState.eval_breaker` to `instrumentation_version` (explained further down). The primary motivation is for performance in free-threaded builds: with thread-local eval breakers, we can stop a specific thread (e.g., for an async exception) without interrupting other threads.

### Pending Calls
We still want to signal all threads in an interpreter for pending calls that aren't limited to the main thread (possible since python/cpython@757b402ea1c2). These callbacks can be added by any thread, with or without the GIL held. If the targeted interpreter's GIL is locked, we signal the holding thread. When any thread is resumed, its `_PY_CALLS_TO_DO` bit is derived from the source of truth for pending calls (one of two `_pending_calls` structs). This handles situations where no thread held the GIL when the call was first added, or if the active thread did not handle the call before releasing the GIL.

In a free-threaded build, all threads in the targeted interpreter are signaled. This is slightly less efficient, both because we have to iterate over all threads to set the bit, and because it's more likely that multiple threads could call into `make_pending_calls()` before the bit is cleared. This shouldn't affect correctness, though, since `make_pending_calls()` and the `_pending_calls` structs have appropriate synchronization in place to ensure each callback happens once.

### Instrumentation Version
The source of truth for the global instrumentation version is still in `PyInterpreterState.instrumentation_version` (which used to be `eval_breaker`), in both normal and free-threaded builds. Now that this variable contains only the instrumentation version (and no more eval breaker bits), some of the code to read and write it is simplified.

Threads usually read the version from their local `eval_breaker`, where it continues to be colocated with the eval breaker bits, and the method for keeping it up to date depends on build type. Both builds first update the version in `instrumentation_version`, and then:
- Normal builds update the version in the current thread's `eval_breaker`. When a new thread takes the GIL, it copies the current version from `instrumentation_version`.
- Free-threaded builds again iterate over all threads in the current interpreter, updating the version on each one. Instrumentation (and the specializing interpreter more generally) will need more work to be compatible with free-threaded builds, so these changes are mostly intended to maintain the status quo in normal builds for now.

### Other notable changes
- The `_PY_*_BIT` macros now expand to the actual bit being set, rather than the bit's index. I think this is simpler overall.
- `_Py_set_eval_breaker_bit(interp, bit, set)` has been split into `_Py_set_eval_breaker_bit(interp, bit)` and `_Py_unset_eval_breaker_bit(interp, bit)`. This lets us use a single atomic `or`/`and`, rather than a loop around an atomic compare/exchange like the old code.

- Issue: gh-112175

### CODE REVIEW

File: Python/instrumentation.c [hunk 1]
```diff
@@ -891,18 +891,39 @@ static inline int most_significant_bit(uint8_t bits) {
...
 }
 
 static void
-set_global_version(PyInterpreterState *interp, uint32_t version)
+set_version_raw(uintptr_t *breaker, uint32_t version)
 {
-    assert((version & _PY_EVAL_EVENTS_MASK) == 0);
-    uintptr_t old = _Py_atomic_load_uintptr(&interp->ceval.eval_breaker);
+    uintptr_t old = _Py_atomic_load_uintptr(breaker);
     intptr_t new;
```
Comment (colesbury) [inline Python/instrumentation.c]: Predates your PR, but this looks like it should be `uintptr_t`:

```suggestion
    uintptr_t new;
```

File: Python/ceval_gil.c [hunk 2]
```diff
@@ -956,38 +1015,38 @@ int
 _Py_HandlePending(PyThreadState *tstate)
 {
     PyInterpreterState *interp = tstate->interp;
```
Comment (colesbury) [inline Python/ceval_gil.c]: variable is unused (see GHA warning)

File: Include/internal/pycore_ceval_state.h [hunk 3]
```diff
@@ -78,11 +78,14 @@ struct _ceval_runtime_state {
...
-     * It is by far the hottest field in this struct and
-     * should be placed at the beginning. */
-    uintptr_t eval_breaker;
+    /* This single variable holds the global instrumentation version and some
+     * interpreter-global requests to break out of the fast path in the eval
+     * loop. PyThreadState also contains an eval_breaker, which is the source
+     * of truth when a thread is running.
+     *
+     * It is by far the hottest field in this struct and should be placed at
+     * the beginning. */
```
Comment (colesbury) [inline Include/internal/pycore_ceval_state.h]: The "hottest field" comment may no longer accurate if `CHECK_EVAL_BREAKER()` now uses the field in PyThreadState.

File: Python/ceval_gil.c [hunk 4]
```diff
@@ -646,11 +661,9 @@ PyEval_RestoreThread(PyThreadState *tstate)
 */
 
 void
-_PyEval_SignalReceived(PyInterpreterState *interp)
+_PyEval_SignalReceived()
```
Comment (colesbury) [inline Python/ceval_gil.c]:
```suggestion
_PyEval_SignalReceived(void)
```

(or take a `_PyRuntimeState *runtime` argument)

File: Python/ceval_gil.c [hunk 5]
```diff
@@ -724,8 +754,16 @@ _PyEval_AddPendingCall(PyInterpreterState *interp,
     int result = _push_pending_call(pending, func, arg, flags);
     PyMutex_Unlock(&pending->mutex);
 
-    /* signal main loop */
-    SIGNAL_PENDING_CALLS(interp);
+    if (main_only) {
+        _PyThreadState_Signal(_PyRuntime.main_tstate, _PY_CALLS_TO_DO_BIT);
+    } else {
```
Comment (colesbury) [inline Python/ceval_gil.c]:
```suggestion
    }
    else {
```

File: Include/internal/pycore_pystate.h [hunk 6]
```diff
@@ -282,6 +282,42 @@ static inline _PyFreeListState* _PyFreeListState_GET(void)
...
+#define _PY_ASYNC_EXCEPTION_BIT (1U << 3)
+#define _PY_GC_SCHEDULED_BIT (1U << 4)
+#define _PY_EVAL_PLEASE_STOP_BIT (1U << 5)
+
+/* Reserve a few bits for future use */
+#define _PY_EVAL_EVENTS_BITS 8
+#define _PY_EVAL_EVENTS_MASK ((1U << _PY_EVAL_EVENTS_BITS)-1)
+
+static inline void
+_PyThreadState_Signal(PyThreadState *tstate, uintptr_t bit)
```
Comment (colesbury) [inline Include/internal/pycore_pystate.h]: I'd suggest keeping the existing naming scheme and placement in `pycore_ceval.h` instead of `pycore_pystate.h` unless there's a concrete reason that requires moving the functions.

I think it still worth splitting setting and clearing the bits like you do here now that we have functions for atomic and & atomic or.

File: Python/ceval_gil.c [hunk 7]
```diff
@@ -702,6 +715,24 @@ _pop_pending_call(struct _pending_calls *pending,
     }
 }
 
+static void
+signal_active_thread(PyInterpreterState *interp, uintptr_t bit)
```
Comment (colesbury) [inline Python/ceval_gil.c:693]: Surround this with an `#ifndef Py_GIL_DISABLED` to address unused function warning

File: Python/gc_free_threading.c [hunk 8]
```diff
@@ -1480,9 +1480,9 @@ PyObject_IS_GC(PyObject *obj)
 }
 
 void
-_Py_ScheduleGC(PyInterpreterState *interp)
+_Py_ScheduleGC(PyThreadState *tstate)
 {
-    _Py_set_eval_breaker_bit(interp, _PY_GC_SCHEDULED_BIT, 1);
+    _PyInterpreterState_SignalAll(tstate->interp, _PY_GC_SCHEDULED_BIT);
```
Comment (colesbury) [inline Python/gc_free_threading.c]: We only need to set the bit for the current thread.

We don't need to worry about missing the signal. If we don't run the GC first, other threads will also see that a GC needs to be scheduled and set the bit on their own eval breakers.

File: Python/instrumentation.c [hunk 9]
```diff
@@ -891,18 +891,39 @@ static inline int most_significant_bit(uint8_t bits) {
...
+{
+    assert((version & _PY_EVAL_EVENTS_MASK) == 0);
+    PyInterpreterState *interp = tstate->interp;
+    set_version_raw(&tstate->interp->ceval.interp_eval_breaker, version);
+
+#ifdef Py_GIL_DISABLED
+    // Set the version on all threads in free-threaded builds.
+    _PyRuntimeState *runtime = &_PyRuntime;
+    HEAD_LOCK(runtime);
+    for (; tstate; tstate = PyThreadState_Next(tstate)) {
```
Comment (colesbury) [inline Python/instrumentation.c]: This looks like it starts iterating from the current thread so it may miss some threads in the interpreter.

I think you will want to start at `interp->threads.head`

File: Python/gc.c [hunk 10]
```diff
@@ -1787,13 +1789,14 @@ _PyObject_GC_Link(PyObject *op)
...
     {
-        _Py_ScheduleGC(tstate->interp);
+        _Py_ScheduleGC(tstate);
     }
 }
 
 void
 _Py_RunGC(PyThreadState *tstate)
 {
+    _PyThreadState_Unsignal(tstate, _PY_GC_SCHEDULED_BIT);
```
Comment (colesbury) [inline Python/gc.c]: I am more comfortable with the bit clearing happening in the places that check for the bit (i.e., `_Py_HandlePending()` and in `signalmodule.c`) rather than being pushed down into this function.

The pattern that I imagine is:

1) Check for the relevant eval breaker bit
2) Clear the bit
3) Take some action

Forgetting to clear the bit leads to problems. It's easier to see that we're adhering to the pattern if checking and clearing the bit are next to each other in, for example `_Py_HandlePending()`.

File: Python/ceval_gil.c [hunk 11]
```diff
@@ -56,60 +56,78 @@
...
+}
 
+// When detaching a thread, transfer _PY_GC_SCHEDULED_BIT to its interpreter,
+// in case a GC was scheduled but not processed yet.
 static inline void
-SIGNAL_PENDING_CALLS(PyInterpreterState *interp)
-{
-    _Py_set_eval_breaker_bit(interp, _PY_CALLS_TO_DO_BIT, 1);
-}
+update_interp_eval_breaker(PyThreadState *tstate, PyInterpreterState *interp) {
```
Comment (colesbury) [inline Python/ceval_gil.c]:
```suggestion
update_interp_eval_breaker(PyThreadState *tstate, PyInterpreterState *interp)
{
```

**Review by colesbury — COMMENTED** [review]
Thanks Brett! This looks good overall, but the free-threaded GitHub CI is failing and there are a few warnings that need to be addressed. I've left comments inline as well.

### _PY_GC_SCHEDULED_BIT

For the free-threaded build, I think we only need to set the GC scheduled bit on the current thread. It doesn't matter much if we get de-scheduled -- other threads will see that a GC needs to be scheduled soon enough. Regarding "multiple threads attempt a GC": that can be a performance issue once the GIL is disabled, but will be addressed in gh-114880.

I think the same logic applies to the default build, but I'd like Mark's opinion on it. In other words, I don't think we need to copy the `_PY_GC_SCHEDULED_BIT` back to the interpreter eval breaker. In the rare case where we get swapped out between setting `_PY_GC_SCHEDULED_BIT` and running the GC, the newly scheduled thread will figure out that it needs to run the GC after the next GC allocation.

### Instrumentation

In the free-threaded build, I think we'll also need to copy the instrumentation bits when we create a new thread (while holding `HEAD_LOCK(runtime)`).

File: Python/gc.c (hunk 10 above)
Comment (swtaarrs) [reply Python/gc.c]: I originally made this change because there were two different places that cleared the bit (`_Py_HandlePending()` and `PyErr_CheckSignals()`), and "clearing the bit" meant different things in normal and free-threaded builds. So I was piggybacking on the existing `gc.c`/`gc_free_threading.c` split to avoid some duplication.

But your reasoning makes sense, and especially with the change to only set the bit on the current thread in free-threaded builds, the behavioral difference goes away.

File: Include/internal/pycore_pystate.h (hunk 6 above)
Comment (swtaarrs) [reply Include/internal/pycore_pystate.h]: I think my only reason for the move was that your 3.12 commit that I used for reference did it (but that was before Mark's gh-109846). I'll move them back unless I come up with something more conrete.

For the naming scheme, how should the functions that operate on all threads in an interpreter fit in? Something like this?
- `_Py_set_eval_breaker_bit(PyThreadState*, uintptr_t)`
- `_Py_unset_eval_breaker_bit(PyThreadState*, uintptr_t)`
- `_Py_set_eval_breaker_bit_all(PyInterpreterState*, uintptr_t)`
- `_Py_unset_eval_breaker_bit_all(PyInterpreterState*, uintptr_t)`

Mildly related: I'd still like to figure out a way to handle thread-agnostic pending calls without signaling all threads, since that will be the only instance of that pattern once I make the GC signaling change. For the main interpreter, I can see it making sense to only signal the main thread (since that's how OS signals are handled), but what about subinterpreters? Would it be reasonable to arbitrarily pick a thread and signal just that one, assuming that any well-behaved Python thread will check its eval_breaker often enough?

File: Include/internal/pycore_ceval_state.h (hunk 3 above)
Comment (swtaarrs) [reply Include/internal/pycore_ceval_state.h]: I'll also remove the padding that says it's there to avoid false sharing. I decided against adding similar padding to `PyThreadState` since it's per-thread rather than per-interpreter, but I can be convinced otherwise.

**Comment by swtaarrs** [comment]
> It doesn't matter much if we get de-scheduled -- other threads will see that a GC needs to be scheduled soon enough.

Thanks, that makes sense. I forgot that every subsequent allocation will keep checking the thresholds even once the GC is scheduled.

> I don't think we need to copy the _PY_GC_SCHEDULED_BIT back to the interpreter eval breaker

If Mark agrees, then the only thing left in `interp_eval_breaker` will be the instrumentation version, and it might be worth renaming it and/or considering larger changes to how it works.

> In the free-threaded build, I think we'll also need to copy the instrumentation bits when we create a new thread (while holding HEAD_LOCK(runtime)).

[This line](https://github.com/python/cpython/pull/115194/files#diff-7ac11e526f79b42d6ea9d3592cb99da46775640c69fa5510f4a6de87cced7141R1309) should do that already in all build types, although I realized that I should change it to read `interp_eval_breaker` atomically. `init_threadstate()` is only called by `new_threadstate()`, and it's called with the runtime lock held.

**Comment by swtaarrs** [comment]
I think I've addressed all of Sam's comments except for removing the `eval_breaker` <-> `interp_eval_breaker` GC bit transfer for normal builds. I'll wait to hear from @markshannon before doing that.

One of the free-threading test failures was the new gc scheduling test, which I deleted in anticipation of removing that functionality. I couldn't repro any of the other failures on either my Linux or MacOS development machines; if the same tests fail again in CI I'll dig further.

File: Include/internal/pycore_ceval.h [hunk 12]
```diff
@@ -200,40 +199,43 @@ int _PyEval_UnpackIterable(PyThreadState *tstate, PyObject *v, int argcnt, int a
...
+#define _PY_CALLS_TO_DO_BIT (1U << 2)
+#define _PY_ASYNC_EXCEPTION_BIT (1U << 3)
+#define _PY_GC_SCHEDULED_BIT (1U << 4)
+#define _PY_EVAL_PLEASE_STOP_BIT (1U << 5)
+#define _PY_EVAL_EXPLICIT_MERGE_BIT (1U << 6)
 
 /* Reserve a few bits for future use */
 #define _PY_EVAL_EVENTS_BITS 8
-#define _PY_EVAL_EVENTS_MASK ((1 << _PY_EVAL_EVENTS_BITS)-1)
+#define _PY_EVAL_EVENTS_MASK ((1U << _PY_EVAL_EVENTS_BITS)-1)
```
Comment (colesbury) [inline Include/internal/pycore_ceval.h]: I think the switch to unsigned here may lead to problems due to zero extension. For example,

```c
uintptr_t mask = ~_PY_EVAL_EVENTS_MASK;  // 0x00000000FFFFFF00
```

where as previously we would have `0xFFFFFFFFFFFFFF00`

**Comment by markshannon** [comment]
> I'll wait to hear from @markshannon before doing that.

What do you want to know?

**Comment by markshannon** [comment]
> If Mark agrees, then the only thing left in interp_eval_breaker will be the instrumentation version, and it might be worth renaming it and/or considering larger changes to how it works.

Feel free to rename it. How did you want to change in how it works?

**Comment by swtaarrs** [comment]
> > I'll wait to hear from @markshannon before doing that.
> 
> What do you want to know?

If you have thoughts on Sam's suggestions around `_PY_GC_SCHEDULED_BIT` in [this comment](https://github.com/python/cpython/pull/115194#pullrequestreview-1872572370). Basically: do we need to worry about a thread dropping the GIL after scheduling a GC, or can we just assume that the next thread that takes the GIL will schedule a GC for itself soon anyway?

> > If Mark agrees, then the only thing left in interp_eval_breaker will be the instrumentation version, and it might be worth renaming it and/or considering larger changes to how it works.
> 
> Feel free to rename it. How did you want to change in how it works?

I didn't have any specific changes in mind, other than making sure we take any opportunities to simplify things once it contains only the version number. We should still keep the version number left-shifted out of the way of the eval breaker bits, since the version is combined with the bits in thread-local eval breakers. Other than that I can't think of anything.

File: Include/internal/pycore_ceval.h (hunk 12 above)
Comment (swtaarrs) [reply Include/internal/pycore_ceval.h]: I think I made this change before I fully paged in how the mask was used; I'll flip it back.

That reminds me that I meant to follow up on [this comment](https://github.com/python/cpython/blob/main/Python/instrumentation.c#L1773-L1775) from @markshannon. That comment, plus other code that operates on the version number as a `uint32_t`, suggest that the upper 32 bits of `eval_breaker` will always be 0 (on platforms where `uintptr_t` is 64 bits). Should we commit to that and make `eval_breaker` a `uint32_t`?

It wouldn't make a meaningful change to the size of the struct but could be worth doing for consistency.

**Comment by markshannon** [comment]
> I think the same logic applies to the default build, but I'd like Mark's opinion on it. In other words, I don't think we need to copy the _PY_GC_SCHEDULED_BIT back to the interpreter eval breaker. In the rare case where we get swapped out between setting _PY_GC_SCHEDULED_BIT and running the GC, the newly scheduled thread will figure out that it needs to run the GC after the next GC allocation.

Correct me if I'm wrong, but we copy the bits from thread state to interpreter state when we are switching threads, and that only happens in the eval breaker handler, which also runs the GC.
So wouldn't `_PY_GC_SCHEDULED_BIT` always be zero when swapping threads as we would have just run the GC and nothing could have triggered another one?

**Comment by colesbury** [comment]
We also can switch threads any time the GIL is released, such as by a `PyEval_SaveThread()` call before a blocking operation.

File: Python/ceval.c [hunk 13]
```diff
@@ -5,7 +5,7 @@
 #include "Python.h"
 #include "pycore_abstract.h"      // _PyIndex_Check()
 #include "pycore_call.h"          // _PyObject_CallNoArgs()
-#include "pycore_ceval.h"         // _PyEval_SignalAsyncExc()
+#include "pycore_ceval.h"
```
Comment (colesbury) [inline Python/ceval.c:8]: Is this change intentional? Otherwise, should probably revert it.

File: Python/ceval_gil.c [hunk 14]
```diff
@@ -56,60 +56,56 @@
...
+#endif
 
-static inline void
-SIGNAL_PENDING_CALLS(PyInterpreterState *interp)
-{
-    _Py_set_eval_breaker_bit(interp, _PY_CALLS_TO_DO_BIT, 1);
-}
+    if (tstate == NULL) {
+        return;
+    }
```
Comment (colesbury) [inline Python/ceval_gil.c]: We should delete this check because `tstate` can't be `NULL` here.

File: Python/instrumentation.c [hunk 15]
```diff
@@ -891,18 +891,42 @@ static inline int most_significant_bit(uint8_t bits) {
 static uint32_t
 global_version(PyInterpreterState *interp)
 {
-    return interp->ceval.eval_breaker & ~_PY_EVAL_EVENTS_MASK;
+    return _Py_atomic_load_uintptr_relaxed(&interp->ceval.instrumentation_version);
```
Comment (colesbury) [inline Python/instrumentation.c]: This needs a cast to `uint32_t` now

**Review by colesbury — COMMENTED** [review]
Looks pretty good, a few comments below

File: Include/internal/pycore_ceval.h (hunk 12 above)
Comment (colesbury) [reply Include/internal/pycore_ceval.h]: Switching to `uint32_t` seems fine, but it also seems fine to keep it as `uintptr_t` in the interest of keeping the PR smaller.

I think previously (i.e., in 3.12) we only had atomic operations implemented for `int` and `uintptr_t`.

File: Python/ceval.c (hunk 13 above)
Comment (swtaarrs) [reply Python/ceval.c:8]: It was intentional: I removed the `_PyEval_SignalAsyncExc()` function and replaced the one call to it (which wasn't even in this file) with `_Py_set_eval_breaker_bit(...)`. I tried removing the `#include` and saw that a number of other functions from `pycore_ceval.h` are still used.

File: Python/instrumentation.c (hunk 15 above)
Comment (swtaarrs) [reply Python/instrumentation.c]: I can add the cast, but I don't think this changed the type of the expression being returned - `eval_breaker` was already a `uintptr_t`, so `interp->ceval.eval_breaker & ~_PY_EVAL_EVENTS_MASK` should also be `uintptr_t`.

File: Include/internal/pycore_ceval_state.h [hunk 16]
```diff
@@ -78,13 +78,10 @@ struct _ceval_runtime_state {
...
-     * the fast path in the eval loop.
-     * It is by far the hottest field in this struct and
-     * should be placed at the beginning. */
-    uintptr_t eval_breaker;
-    /* Avoid false sharing */
-    int64_t padding[7];
+    /* This variable holds the global instrumentation version. When a thread is
+       running, this value is overlaid onto PyThreadState.eval_breaker so that
+       changes in the instrumentation version will trigger the eval breaker. */
+    uintptr_t instrumentation_version;
```
Comment (colesbury) [inline Include/internal/pycore_ceval_state.h:84]: Now that this is just the instrumentation_version, did you want to make this a `uint32_t` and simplify `set_version_raw`?

**Review by colesbury — COMMENTED** [review]
This looks good to me other than the merge conflicts. Those were probably from conflicts with my PRs -- sorry about that.

Did you want to make `instrumentation_version` a `uint32_t` now that it doesn't include the eval breaker? Either way seems fine to me.

File: Include/internal/pycore_ceval_state.h (hunk 16 above)
Comment (swtaarrs) [reply Include/internal/pycore_ceval_state.h:84]: I decided to keep it as-is to keep the PR more focused. To keep things more consistent it would make sense to change the new `eval_breaker` to `uint32_t` as well, and that would make the change significantly bigger.

**Comment by swtaarrs** [comment]
I resolved the merge conflicts - they were pretty straightforward. This should be good to go now (I replied inline about not making the `uint32_t` change in this PR).

**Review by colesbury — APPROVED** [review]
Thanks Brett!
